File: rust/src/models/add_on.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{
    add_on_id::AddOnId, entitlement::Entitlement, price_id::PriceId,
    product_fee_type_enum::ProductFeeTypeEnum, product_id::ProductId,
};

#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
pub struct AddOn {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub archived_at: Option<String>,

    pub created_at: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub entitlements: Option<Vec<Entitlement>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub fee_type: Option<ProductFeeTypeEnum>,

    pub id: AddOnId,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_instances_per_subscription: Option<i32>,

    pub name: String,

    pub price_id: PriceId,

    pub product_id: ProductId,

    pub self_serviceable: bool,
}

impl AddOn {
    pub fn new(
        created_at: String,
        id: AddOnId,
        name: String,
        price_id: PriceId,
        product_id: ProductId,
        self_serviceable: bool,
    ) -> Self {
        Self {
            archived_at: None,
            created_at,
            description: None,
            entitlements: None,
            fee_type: None,
            id,
            max_instances_per_subscription: None,
            name,
            price_id,
            product_id,
            self_serviceable,
        }
    }
}

File: typescript/src/models/addOn.ts
// this file is @generated
import { type AddOnId, AddOnIdSerializer } from "./addOnId";
import { type Entitlement, EntitlementSerializer } from "./entitlement";
import { type PriceId, PriceIdSerializer } from "./priceId";
import {
  type ProductFeeTypeEnum,
  ProductFeeTypeEnumSerializer,
} from "./productFeeTypeEnum";
import { type ProductId, ProductIdSerializer } from "./productId";

export interface AddOn {
  archivedAt?: Date | null;

  createdAt: Date;

  description?: string | null;

  entitlements?: Entitlement[];

  feeType?: ProductFeeTypeEnum | null;

  id: AddOnId;

  maxInstancesPerSubscription?: number | null;

  name: string;

  priceId: PriceId;

  productId: ProductId;

  selfServiceable: boolean;
}

export const AddOnSerializer = {
  _fromJsonObject(object: any): AddOn {
    return {
      archivedAt:
        object["archived_at"] != null ? new Date(object["archived_at"]) : undefined,
      createdAt: new Date(object["created_at"]),
      description: object["description"],
      entitlements:
        object["entitlements"] != null
          ? object["entitlements"].map((item: any) =>
              EntitlementSerializer._fromJsonObject(item)
            )
          : undefined,
      feeType:
        object["fee_type"] != null
          ? ProductFeeTypeEnumSerializer._fromJsonObject(object["fee_type"])
          : undefined,
      id: AddOnIdSerializer._fromJsonObject(object["id"]),
      maxInstancesPerSubscription: object["max_instances_per_subscription"],
      name: object["name"],
      priceId: PriceIdSerializer._fromJsonObject(object["price_id"]),
      productId: ProductIdSerializer._fromJsonObject(object["product_id"]),
      selfServiceable: object["self_serviceable"],
    };
  },

  _toJsonObject(self: AddOn): any {
    return {
      archived_at: self.archivedAt,
      created_at: self.createdAt,
      description: self.description,
      entitlements:
        self.entitlements != null
          ? self.entitlements.map((item: any) =>
              EntitlementSerializer._toJsonObject(item)
            )
          : undefined,
      fee_type:
        self.feeType != null
          ? ProductFeeTypeEnumSerializer._toJsonObject(self.feeType)
          : undefined,
      id: AddOnIdSerializer._toJsonObject(self.id),
      max_instances_per_subscription: self.maxInstancesPerSubscription,
      name: self.name,
      price_id: PriceIdSerializer._toJsonObject(self.priceId),
      product_id: ProductIdSerializer._toJsonObject(self.productId),
      self_serviceable: self.selfServiceable,
    };
  },
};
